intel_extension_for_transformers/llm/runtime/graph/fused_attention.md

```diff
@@ -0,0 +1,24 @@
+Fused Attention
+=======
+
+Attention (including MHA, GQA, MHA) is one of the key parts of transformers and also the bottleneck in many scenarios. To implement various optimizations, a fused attention layer and corresponding utilities for the customized KV-cache it uses are introduced.
+
+Note, this doc assumes you have the basic knowledge of this cpp graph implementation including `ne_tensor::ne`, `ne_tensor::nb` etc.
+
+## KV-cache initialization
+The memory for kv-cache is allocated in `/models/model_utils/model_utils.cpp`. As the fused attention implementation requires certain instruction extensions and potentially some other limitations, fused attention is enabled only if `jblas_reordered_attn_fp32_support()` aggress, denoting with `memory_type = NE_TYPE_JBLAS`. Next, `get_batch_kv_elements_from_gpt_params()` will give the sizes (in terms of bytes if fused attention enabled, or in terms elements if fused attention disabled) of k-cache and v-cache respectively for each batch and each layer. The KV-cache is finally prepared with these 2 sizes by creating `ne_new_tensor` inside `model.kv_self.k/.v` or `model.layer[il].k_cache/.v_cache`.
+
+## KV-cache Append
+KV-cache is appended every time a new pair of K and V are generated by evaluating inner product for QKV (ne_mul_qkv). This operation append an additional K/V-tensor on the dimension of `n_past` (or the dimension of sequence length) (i.e. resulting `n_past = n_past + N`).
```

Comment:
you can give the code snippet

Reply:
Note sure if putting code for KV-cache updating here would help. For example, `ne_view_4d` and `ne_build_forward_expand` is hard to understand for readers who are not familiar with cpp graph code (or ggml code).